Restyling of search bar

hamburge opened the pull request:
Search bar styling to match the new designs. There are some missing svgs--rectangles currently replace all of the missing item type svgs, the circle on the right will be an X to clear the inputted value, and the magnifying glass will be a different version that is flipped the other direction.

Other changes:
- Show all possible filters, not just if they have corresponding options available. For example, if there are no Smart Objects in the document, still show the Smart Objects filter option.
- Fix broken datalist updating- the filter icon would not always show until the render after selecting the filter.

The designs:
![design-space-search-001](https://cloud.githubusercontent.com/assets/12618145/8995981/cf9b097c-36c9-11e5-84be-0bcd5e0e1d13.png)
![design-space-search-002](https://cloud.githubusercontent.com/assets/12618145/8995982/cf9e01ea-36c9-11e5-859f-2f1b9ba5e792.png)
![design-space-search-003](https://cloud.githubusercontent.com/assets/12618145/8995983/cf9e97a4-36c9-11e5-92c1-b893550d1c1e.png)
![application-search-002](https://cloud.githubusercontent.com/assets/12618145/8995991/d7cb1178-36c9-11e5-8303-756bc235362b.png)
![application-search-005](https://cloud.githubusercontent.com/assets/12618145/8996151/c7dad52c-36ca-11e5-887f-100cb9dc93a3.png)

iwehrman commented:
The search bar does not seem to be centered in the non-UI bounds:
![center](https://s3.amazonaws.com/f.cl.ly/items/0V0y140n013h1p3R0U07/Image%202015-07-30%20at%203.25.29%20PM.png)

iwehrman commented:
The inner hairline border is too light/bright compared to the comp:
![border](https://s3.amazonaws.com/f.cl.ly/items/312y2s3i2b201F1M3E1Y/Image%202015-07-30%20at%203.26.53%20PM.png)

iwehrman commented:
All the filter icons are missing:
![filter](https://s3.amazonaws.com/f.cl.ly/items/2N1Z1p071Y2o253Y2O1W/Image%202015-07-30%20at%203.28.14%20PM.png)

hamburge commented:
I fixed the colors and icons, but I'm having trouble centering it based on the new layout. Any tips @designedbyscience?

Also for now I changed the maximum number of options back to a smaller number, because there's an issue with scrolling, so I'll put that in another PR.

iwehrman commented:
This probably doesn't help, but it looks like the inner div is bleeding off the left edge of the correctly centered dialog element:
![dialog](https://s3.amazonaws.com/f.cl.ly/items/1f2c3w032t1g200l2t2B/Image%202015-07-30%20at%206.28.46%20PM.png)

iwehrman commented:
Also, the clear-search icon is currently a circle. Are you in need of an X icon? How about `&times;`?

hamburge commented:
-Fixed scrolling so now we can have more options
-Replaced the circle with an X to clear (shoutout to unicode and @iwehrman)
-Placeholder text now tells you what you are searching for if you have a filter selected
-If your item doesn't have a name it now places "Untitled" in italics as its name
-Adjusted styling for filter options after chatting with Seth

Changes abound!

iwehrman commented:
But the centering!?!

iwehrman commented on src/js/stores/search.js:
```diff
@@ -231,18 +239,26 @@ define(function (require, exports, module) {
 
+                    if (name === "") {
+                        name = "Untitled";
+                        style = { "font-style": "italic" };
```
This should be `fontStyle`: https://facebook.github.io/react/tips/inline-styles.html

iwehrman commented on src/style/search/search-bar.less:
```diff
@@ -85,38 +98,68 @@
+
+.search-bar__dialog .select__header {
+    color: @search-border;
+    border-top: 1px solid @search-border;
```
Also, you can't use pixel constants in the CSS. Everything has to be in rems.

iwehrman commented on src/js/stores/search.js:
```diff
@@ -200,16 +209,15 @@ define(function (require, exports, module) {
+                        var options = this._getOptions(type);
+
+                        return items.push(options);
+                    }.bind(this), Immutable.List());
```
It would be better to build this with an array and then convert the complete array to an immutable list at the end.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -379,17 +330,22 @@ define(function (require, exports, module) {
+                        title="Clear Search Input"
+                        className="button-clear-search"
+                        onClick={this._clearInput}
+                       >
```
This delimiter should go back up at the end of the previous line.

iwehrman commented:
I give this PR :hamburger: :hamburger: :hamburger: :hamburger: :hamburger:
